commons-statistics-distribution/src/main/java/org/apache/commons/statistics/distribution/TruncatedNormalDistribution.java

```diff
@@ -0,0 +1,206 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.commons.statistics.distribution;
+
+/**
+ * Implementation of the truncated normal distribution.
+ *
+ * @see <a href="http://en.wikipedia.org/wiki/Truncated_normal_distribution">
+ * Truncated normal distribution (Wikipedia)</a>.
+ */
+public class TruncatedNormalDistribution extends AbstractContinuousDistribution {
+    /**
+     * Mean of parent normal distribution.
+     */
+    private final double parentMean;
+    /**
+     * Standard deviation of parent normal distribution.
+     */
+    private final double parentSd;
+    /**
+     * Mean of this distribution.
+     */
+    private final double mean;
+    /**
+     * Variance of this distribution.
+     */
+    private final double variance;
+    /**
+     * Lower bound of this distribution.
+     */
+    private final double lower;
+    /**
+     * Upper bound of this distribution.
+     */
+    private final double upper;
+
+    /**
+     * A standard normal distribution used for calculations.
+     */
+    private final NormalDistribution standardNormal;
+    /**
+     * Stored value of @{code standardNormal.cumulativeProbability((lower - mean) / sd)} for faster computations.
+     */
+    private final double cdfAlpha;
+    /**
+     * Stored value of @{code (upper - mean) / sd) - cdfAlpha} for faster computations.
+     */
+    private final double cdfDelta;
+
+    /**
+     * Creates a truncated normal distribution.
+     * Note that the {@code mean} and {@code sd} is of the parent normal distribution,
+     * and not the true mean and standard deviation of the truncated normal distribution.
+     *
+     * @param mean mean for this distribution.
+     * @param sd standard deviation for this distribution.
+     * @param lower lower bound (inclusive) of the distribution, can be {@link Double#NEGATIVE_INFINITY}.
+     * @param upper upper bound (inclusive) of the distribution, can be {@link Double#POSITIVE_INFINITY}.
+     * @throws IllegalArgumentException if {@code sd <= 0} or if {@code upper <= lower}.
+     */
+    public TruncatedNormalDistribution(double mean, double sd, double lower, double upper) {
+        if (sd <= 0) {
+            throw new DistributionException(DistributionException.NEGATIVE, sd);
```

Comment:
I am thinking that we should have a `DistributionException.NOT_STRICTLY_POSITIVE` constant here. However the same message is used in NormalDistribution (and others) for zero so if you introduce this new constant then I can propagate it where appropriate through the rest of the code. Or you can leave it and I can do this after this PR is merged.